doc/reference/configuration/configuration_reference.rst

```diff
@@ -3824,6 +3824,219 @@ snapshot.by.*
+
+    The cache size for the vinyl storage engine. The cache can
+    be resized dynamically.
+
+    |
+    | Type: integer
+    | Default: 128 * 1024 * 1024
+    | Environment variable: TT_VINYL_CACHE
+
+..  _configuration_reference_vinyl_defer_deletes:
+
+..  confval:: vinyl.defer_deletes
+
+    Enable the deferred DELETE optimization in vinyl. It was disabled by default
+    since Tarantool version 2.10 to avoid possible performance degradation
+    of secondary index reads.
+
+    |
+    | Type: boolean
+    | Default: false
+    | Environment variable: TT_VINYL_DEFER_DELETES
+
+..  _configuration_reference_vinyl_dir:
+
+..  confval:: vinyl.dir
+
+    A directory where vinyl files or subdirectories will be stored.
+
+    This option may contain a relative file path.
+    In this case, it is interpreted as relative to
+    :ref:`process.work_dir <configuration_reference_process_work_dir>`.
+
+    |
+    | Type: string
+    | Default: var/lib/{{ instance_name }}'
+    | Environment variable: TT_VINYL_DIR
+
+..  _configuration_reference_vinyl_max_tuple_size:
+
+..  confval:: vinyl.max_tuple_size
+
+    Size of the largest allocation unit,
+    for the vinyl storage engine. It can be increased if it
+    is necessary to store large tuples.
+    See also: :ref:`memtx.max_tuple_size <configuration_reference_memtx_max_tuple_size>`.
+
+    |
+    | Type: integer
+    | Default: 1024 * 1024
+    | Environment variable: TT_VINYL_MAX_TUPLE_SIZE
+
+..  _configuration_reference_vinyl_memory:
+
+..  confval:: vinyl.memory
+
+    The maximum number of in-memory bytes that vinyl uses.
+
+    |
+    | Type: integer
+    | Default: 28 * 1024 * 1024
+    | Environment variable: TT_VINYL_MEMORY
+
+..  _configuration_reference_vinyl_page_size:
+
+..  confval:: vinyl.page_size
+
+    Page size. Page is a read/write unit for vinyl disk operations.
+    The ``vinyl.page_size`` setting is a default value for one of the
+    options in the :ref:`Options for space_object:create_index() <box_space-create_index>` chart.
+
+    |
+    | Type: integer
+    | Default: 8 * 1024
+    | Environment variable: TT_VINYL_PAGE_SIZE
+
+..  _configuration_reference_vinyl_range_size:
+
+..  confval:: vinyl.range_size
+
+    The default maximum range size for a vinyl index, in bytes.
+    The maximum range size affects the decision whether to
+    :ref:`split <engines-vinyl_split>` a range.
+
+    If ``vinyl.range_size`` is not nil and not 0, then
+    it is used as the
+    default value for the ``range_size`` option in the
+    :ref:`Options for space_object:create_index() <box_space-create_index>` chart.
+
+    If ``vinyl.range_size`` is nil or 0, and ``range_size`` is not specified
+    when the index is created, then Tarantool sets a value later depending on
+    performance considerations. To see the actual value, use
+    :doc:`index_object:stat().range_size </reference/reference_lua/box_index/stat>`.
+
+    |
+    | Type: integer
+    | Default: box.NULL
```

Comment:
A reader can guess that a default behavior is described in this 'nil or 0' paragraph. However, the default is `box.NULL` and technically such a case is not described in the option description.

Moreover, a reader likely works with this machinery by creating/editing a YAML file, where we have neither `nil`, nor `box.NULL` (but we have a missed value and `null`).

I think we should make a little stop here and look how the options, which has `box_cfg` annotation in the schema, works.

----

First, a YAML file (or YAML document from a config storage) is parsed into Lua. If there is no option in YAML, it is `nil` in Lua. If the option is `null` in YAML, it is `box.NULL` in Lua.

| config.yaml | no option | `null`     |
| ----------- | --------- | ---------- |
| Lua         | `nil`     | `box.NULL` |

Next, default values are applied. Let's assume that the `default` annotation contains `box.NULL`.

| config.yaml   | no option  | `null`     |
| ----------- | --------- | ---------- |
| Lua           | `nil`      | `box.NULL` |
| apply default | `box.NULL` | `box.NULL` |

The important fact here is that a missing option is the same as `null` option.

Then the option is passed to the `box.cfg()` call. The `box.cfg()` code assumes that `box.NULL` and `''` (an empty string) means a default value (from the `default_cfg` table in `src/box/lua/load_cfg.lua`).

The important fact here is that `default = box.NULL` for an option with the `box_cfg` annotation means that the effective default is determined somewhere under the hood.

Usually `instance_config.lua` has `box.NULL` default for options, whose default is either hard to determine from Lua (depends on the OS, for example) or if the default is determined in runtime.

`load_cfg.lua::default_cfg.vinyl_range_size` is `nil`. Actually, it means that the effective default is determined in runtime.

Going down by the code we can see that `vinyl_range_size = nil` is represented as 0 in the C code and it means that a missed option, `null` and 0 in YAML are equivalent.

Going further, we can see that if the option is 0 in C, some heuristic is applied (`vy_lsm_range_size` in `src/box/vy_lsm.c`).

----

To sum up:

1. `box.NULL` is a marker that means that the effective default is determined in runtime. `config:get()` returns `box.NULL`. An effective value is something that `vy_lsm_range_size()` returns.

   It is a bit different from the existing `Default: box.NULL`, as for me. I would describe it somehow like `Default: box.NULL (means that an effective default is determined in runtime)`.
3. Missed option, `null` and 0 are equivalent in the case.
4. `nil` may confuse a reader of this document, because the document is likely read in context of YAML.

Reply:
Thanks a lot for the details! My suggestion, please say if it's ok:

```
    <...>
    If ``vinyl.range_size`` is specified (but the value is not ``null`` or 0), then
    it is used as the default value for the ``range_size`` option in the
    :ref:`Options for space_object:create_index() <box_space-create_index>` chart.

    If ``vinyl.range_size`` is not specified (or is explicitly set to ``null`` or 0),
    and ``range_size`` is not specified when the index is created,
    then Tarantool sets a value later depending on performance considerations.
    To see the actual value, use
    :doc:`index_object:stat().range_size </reference/reference_lua/box_index/stat>`.

    |
    | Type: integer
    | Default: box.NULL (means that an effective default is determined in runtime)
    | Environment variable: TT_VINYL_RANGE_SIZE
```

Reply:
Seems OK, thanks for the update!